### PR TITLE
Implemented ShooterCommand, bound it to buttons

### PR DESCRIPTION
**Pull request by Weq511762338**

### CODE REVIEW

**Comment by coachFitz0807** [comment]
Pretty good. Just check Vision status before reading the distance.

File: src/main/java/frc/robot/Constants.java [hunk 1]
```diff
@@ -77,12 +77,22 @@
...
     public static final double SHOOTER_KF = 1023.0/20660.0;  // CTRE example = 1023.0/20660.0
 
+    // amount of time to wait for the motors on shooters to speed up, later replaced with checking RPM
+    public static final double SHOOTER_MOTOR_WAIT_TIME = 0.5;
+    // time to wait for the shots from the shooter
+    public static final double SHOOTER_INTAKE_WAIT_TIME = 0.25;
+    // amount of time after spin-up to do 2 shots
+    public static final double SHOTS_WAIT_TIME = 0.5;
+    // amount of time for the intake to carry the next ball to the chute
+    public static final double INTAKE_WAIT_TIME = 0.5;
```
Comment (prensing) [inline src/main/java/frc/robot/Constants.java:87]: SHOTS_WAIT_TIME and INTAKE_WAIT_TIME are poor names. INTAKE_WAIT_TIME does not even mention shooting.
Maybe use "SHOOTER_BALL1_WAIT_TIME" and "SHOOTER_BALL2_WAIT_TIME"?

File: src/main/java/frc/robot/commands/ShooterCommand.java [hunk 2]
```diff
@@ -1,42 +1,105 @@
...
         m_shooter = shooter;
         m_intake = intake;
         m_vision = vision;
     }
 
+    // Called when the command is initially scheduled.
     @Override
     public void initialize() {
-        System.out.println("*sound of shooter shooting*"); // debugging
+        m_state = State.SPEED_UP_SHOOTER;
```
Comment (prensing) [inline src/main/java/frc/robot/commands/ShooterCommand.java:43]: This is the wrong start state.

File: src/main/java/frc/robot/commands/ShooterCommand.java [hunk 3]
```diff
@@ -1,42 +1,105 @@
...
     // Called every time the scheduler runs while the command is scheduled.
     @Override
     public void execute() {
-        
+        switch (m_state) {
+            case FINDING_VISION_TARGET:
+                m_distance = m_vision.getDistance();
+                if(m_distance != 0.0)
+                    m_state = State.SPEED_UP_SHOOTER;
+                break;
```
Comment (prensing) [inline src/main/java/frc/robot/commands/ShooterCommand.java:54]: This should be a fall-through. make sure to include a comment.

File: src/main/java/frc/robot/commands/ShooterCommand.java [hunk 4]
```diff
@@ -1,42 +1,105 @@
...
+                    m_state = State.TURN_ON_INTAKE;
+                break;
+            
+            case TURN_ON_INTAKE:
+                m_intake.run(Constants.INTAKE_SHOOTING_SPEED);
+                m_intakeTime.start();
+                m_state = State.WAIT_FOR_INTAKE;
+                break;
+
+            // State.WAIT_FOR_INTAKE checked in isFinished method
```
Comment (prensing) [inline src/main/java/frc/robot/commands/ShooterCommand.java:88]: This should be an actual case statement. That will get rid of the warning at the "switch" line.

File: src/main/java/frc/robot/Constants.java (hunk 1 above)
Comment (prensing) [reply src/main/java/frc/robot/Constants.java:87]: Also need to update the comments to match the usage.

File: src/main/java/frc/robot/commands/ShooterCommand.java (hunk 3 above)
Comment (prensing) [reply src/main/java/frc/robot/commands/ShooterCommand.java:54]: Unfortunately, I think we want a timeout on this state. The vision might not be working and we don't want to wait forever. I think you can use a LigerTimer with maybe 2 seconds, but it does not need to be a Constants number. If it times out, use 9 ft for the distance (should be local constant in the class).

**Review by prensing — CHANGES_REQUESTED** [review]

**Review by prensing — APPROVED** [review]

**Pull request merged**